### PR TITLE
Spells recharge by INVOKE_ASYNC() instead of by processing

### LINKED ISSUE #27391: Xeno tail sweep button does not turn red when in cooldown.
What it says on the tin.

### PR DESCRIPTION
**Pull request by ChangelingRain**
on the list of things that didn't need to and shouldn't have been done, this.
just use INVOKE_ASYNC dude
Fixes #27391

### CODE REVIEW

**Comment by coiax** [comment]
Why not just use timers?

**Comment by ChangelingRain** [comment]
You can't easily display the time until it finishes then, can you?

**Comment by coiax** [comment]
I'm not sure if there's a proc for getting time til the timer fires with the timer's ID. @MrStonedOne?

**Comment by MrStonedOne** [comment]
There is a per thing cost with the byond sleep/spawn queue for every thing inserted into it because adding to the sleep/spawn queue is O(n) since byond does a sorted insert. 

So having too many items in the sleep queue slows down all sleep operations.

If we aren't going to make spells burstable (where they can store more charge then they need to fire, something spells don't do right now iirc) then why store the charge at all? just store the time it was last used and calculate it on the fly

**Comment by ChangelingRain** [comment]
spells have the stat panel thing where you can see the charge, and they need some sort of callback to the action once they return to full charge so the action can become visibly 'usable'.

**Comment by Cyberboss** [comment]
Yeah, I was gonna suggest this change to @kevinz000 but realistically it's better as process

**Comment by MrStonedOne** [comment]
Then it might be better to do this in it's own SubSystem.

While this may be an extreme example, it shows how badly the byond sleep system scales, and why i've supported moving away from it to timers or excessive subsystems.

Lets say there are 75 people playing, and the wiznerd uses summon magic four times, lets just say that on avg people will get 3 spells and 1 staff (ish).

In the worst case, if all of those spells are recharging because the players just started going wild, every time one spell goes to sleep while recharging, byond will have to go step thru 224 other sleeping procs in the sleep queue linked list to find the position for this.

Every decisecond, 225 spells each have to filter thru the 224 other recharging spells that exist in the sleep queue. So thats 50,400 comparisons, every dedi second.

IE: Per thing sleeping loops suck balls, and merging them up to a system is generally better for performance.

File: code/modules/spells/spell.dm [hunk 1]
```diff
@@ -251,24 +250,19 @@ GLOBAL_LIST_INIT(spells, typesof(/obj/effect/proc_holder/spell)) //needed for th
 /obj/effect/proc_holder/spell/proc/start_recharge()
 	if(action)
 		action.UpdateButtonIcon()
-	recharging = TRUE
+	while(charge_counter < charge_max && !QDELETED(src))
+		sleep(1)
```
Comment (Iamgoofball) [inline code/modules/spells/spell.dm]: >>>>

**Comment by kevinz000** [comment]
Dangit Joan I was told no sleeps ever :(

**Comment by optimumtact** [comment]
it's better that their processing for all the reasons mso mentioned

**Comment by Cyberboss** [comment]
Just close it then

**Pull request closed without merging**